### PR TITLE
Remove `get_aiida_class`

### LINKED ISSUE #2237: Reorganise .get_aiida_class()
In the new provenance design:
- We should not have at the Model classes a method that creates an ORM instance given a Model instance.
- We should have at the backend abstraction layer an method/function that creates a corresponding backend entity to the given models entity. Then at the ORM level we should have the generation of a corresponding AiiDA ORM entity to the given backend entity.

### PR DESCRIPTION
**Pull request by muhrin**
Fixes #2237 

This is part of the procedure that @szoupanos is leading to convert to automatically generated dummy models for the `QueryBuilder`

Replaced all instances with either get_backend_entity or get_orm_entity
(as appropriate).

Also changed backend group to be independent of the ORM.  Because the
iterator was getting the orm class and returning that (which required it
to access up to the ORM level).  Now it just returns backend entities
which are in turn converted to ORM entities by the orm.Group.  This also
means you can't call len(group.nodes) because this is not allowed for
iterators but the same can be achieved with `len(list(group.nodes))`.

Removal of `get_aiida_class` from QueryBuilder and addition of
`DbModel-to-BackendEntity` and `BackendEntity-to-OrmEntity` convertors.

### CODE REVIEW

File: aiida/orm/implementation/sqlalchemy/querybuilder.py [hunk 1]
```diff
@@ -27,6 +27,10 @@
 from aiida.common.exceptions import InputValidationError
 from aiida.orm.implementation.querybuilder import BackendQueryBuilder
 
+from aiida.orm.implementation.sqlalchemy.convert import get_backend_entity
+
+# from aiida.orm.factories import EntityFactory
```
Comment (sphuber) [inline aiida/orm/implementation/sqlalchemy/querybuilder.py]: dead code

File: aiida/orm/implementation/sqlalchemy/workflow.py [hunk 2]
```diff
@@ -839,10 +842,11 @@ def get_workflow_info(w, tab_size=2, short=False, pre_string="",
         #     warnings.simplefilter("ignore", category=sa_exc.SAWarning)
```
Comment (sphuber) [inline aiida/orm/implementation/sqlalchemy/workflow.py]: dead code

File: aiida/orm/querybuilder.py [hunk 3]
```diff
@@ -1942,6 +1947,8 @@ def first(self):
         try:
             returnval = [
                 self._impl.get_aiida_res(self._attrkeys_as_in_sql_result[colindex], rowitem)
+                # factories.EntityFactory.get_orm_from_backend_entity(
+                #     self._impl.get_backend_entity_res(self._attrkeys_as_in_sql_result[colindex], rowitem))
```
Comment (sphuber) [inline aiida/orm/querybuilder.py]: dead code

File: aiida/orm/querybuilder.py [hunk 4]
```diff
@@ -1952,6 +1959,8 @@ def first(self):
             # It still returns a list!
             else:
                 returnval = [self._impl.get_aiida_res(self._attrkeys_as_in_sql_result[0], resultrow)]
+                # returnval = [factories.EntityFactory.get_orm_from_backend_entity(
+                #     self._impl.get_backend_entity_res(self._attrkeys_as_in_sql_result[0], resultrow))]
```
Comment (sphuber) [inline aiida/orm/querybuilder.py]: dead code

File: aiida/orm/implementation/django/groups.py [hunk 5]
```diff
@@ -159,12 +162,12 @@ def __next__(self):
             def next(self):
                 return next(self.generator)
 
-        return Iterator(self._dbmodel.dbnodes.all())
```
Comment (giovannipizzi) [inline aiida/orm/implementation/django/groups.py]: Better to replace this with `self._dbmodel.dbnodes.all().iterator()` for lazy loading also while looping with __iter__ and __next__, see
see also here http://blog.etianen.com/blog/2013/06/08/django-querysets/

File: aiida/orm/groups.py [hunk 6]
```diff
@@ -269,7 +273,7 @@ def nodes(self):
         the respective AiiDA subclasses of Node, and also allows to ask for
         the number of nodes in the group using len().
         """
-        return self._backend_entity.nodes
+        return map(convert.get_orm_entity, self._backend_entity.nodes)
```
Comment (giovannipizzi) [inline aiida/orm/groups.py]: We should use a custom Iterator class as it is done in the Backend classes to have __len__ pipe through the .count() method of the ORM, for efficiency

File: aiida/cmdline/commands/cmd_group.py [hunk 7]
```diff
@@ -79,7 +79,7 @@ def group_delete(group, force):
     group_pk = group.pk
     group_name = group.name
 
-    num_nodes = len(group.nodes)
```
Comment (giovannipizzi) [inline aiida/cmdline/commands/cmd_group.py]: This should go back to `len(group.nodes)` once the suggested fix for __len__ is in place, for efficiency

File: aiida/orm/groups.py (hunk 6 above)
Comment (muhrin) [reply aiida/orm/groups.py]: Done

File: aiida/cmdline/commands/cmd_group.py (hunk 7 above)
Comment (muhrin) [reply aiida/cmdline/commands/cmd_group.py]: Done.

**Pull request merged**